res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,3179 @@
+        }
+        if (this.unshift !== undefined && typeof this.unshift === "function") {
+            this.unshift();
+        }
+        this.shifted = false;
+        if (this.input !== undefined && typeof this.input === "function"
+            && this.inReport !== undefined && this.inReport instanceof HIDInputReport) {
+            this.inConnect();
+        }
+        this.outConnect();
+    }
+    inConnect(callback) {
+        if (this.inByte === undefined
+            || this.inBit === undefined
+            || this.inBitLength === undefined
+            || this.inReport === undefined) {
+            return;
+        }
+        if (typeof callback === "function") {
+            this.input = callback;
+        }
+        this.inConnection = this.inReport.registerCallback(this.input.bind(this), this.inByte, this.inBit, this.inBitLength, this.oldDataDefault);
+    }
+    inDisconnect() {
+        if (this.inConnection !== undefined) {
+            this.inConnection.disconnect();
+        }
+    }
+    send(value) {
+        if (this.outReport !== undefined && this.outByte !== undefined) {
+            this.outReport.data[this.outByte] = value;
+            this.outReport.send();
+        }
+    }
+    output(value) {
+        this.send(value);
+    }
+    outConnect() {
+        if (this.outKey !== undefined && this.group !== undefined) {
+            const connection = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            // This is useful for case where effect would have been fully disabled in Mixxx. This appears to be the case during unit tests.
+            if (connection) {
+                this.outConnections[0] = connection;
+            } else {
+                console.warn("Unable to connect '" + this.group + "." + this.outKey + "' to the controller output. The control appears to be unavailable.");
+            }
+        }
+    }
+    outDisconnect() {
+        for (const connection of this.outConnections) {
+            connection.disconnect();
+        }
+        this.outConnections = [];
+    }
+    outTrigger() {
+        for (const connection of this.outConnections) {
+            connection.trigger();
+        }
+    }
+}
+class ComponentContainer extends Component {
+    constructor() {
+        super();
+    }
+    *[Symbol.iterator]() {
+        // can't use for...of here because it would create an infinite loop
+        for (const property in this) {
+            if (Object.prototype.hasOwnProperty.call(this, property)) {
+                const obj = this[property];
+                if (obj instanceof Component) {
+                    yield obj;
+                } else if (Array.isArray(obj)) {
+                    for (const objectInArray of obj) {
+                        if (objectInArray instanceof Component) {
+                            yield objectInArray;
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
It is recursive. 

Line 377, the for loop iterate over `obj`'s `*[Symbol.iterator]()`

Reply:
Thats only the case if that `obj` is an array (L376). And even then, since `ComponentContainer` is inheriting `Component`, the first case (L374) will be taken instead and it will just `yield obj;`

Reply:
Actually, sorry, it's not directly recursive for a reason.

The point of this implementation was to flatten every direct children `Component` so then can be `(un)shifted` by a `ComponentContainer`, which itself override `(un)shift` to apply this logic. So this was on purpose that this function doesn't iterate through children's containers.

The point of doing this is to allow a `(Component)Container` to potentially stop the propagation. Perhaps it would have been better to use an explicit method (e.g `children()`) rather than overriding the `Symbol.iterator`, but this code was already there so I've simply allow resistivity with the `shift/unshift` couple. Happy to rename if you thing it's confusing.

Reply:
Ah, that makes total sense. I also prefer to make that behavior customizable (I had to work around it a couple times now). The only problem I see is that it makes the assumption that the caller implicitly continues the recursion I think we could come up with a better design for that in the future.